test/integration/document_test.ts

```diff
@@ -761,6 +762,163 @@ describe('Document', function () {
+    const unsub1 = d1.subscribe('status', (event) => {
+      eventCollectorD1.add(event.value);
+    });
+    const unsub2 = d2.subscribe('status', (event) => {
+      eventCollectorD2.add(event.value);
+    });
+
+    // 1. When the client attaches a document, it receives an attached event.
+    await c1.attach(d1);
+    await c2.attach(d2);
+
+    await eventCollectorD1.waitAndVerifyNthEvent(1, {
+      status: DocumentStatus.Attached,
+      actorID: c1ID,
+    });
+    await eventCollectorD2.waitAndVerifyNthEvent(1, {
+      status: DocumentStatus.Attached,
+      actorID: c2ID,
+    });
+
+    // 2. When c1 detaches a document, it receives a detached event.
+    await c1.detach(d1);
+    await eventCollectorD1.waitAndVerifyNthEvent(2, {
+      status: DocumentStatus.Detached,
+    });
+
+    // 3. When c2 deactivates, it should also receive a detached event.
+    await c2.deactivate();
+    await eventCollectorD2.waitAndVerifyNthEvent(2, {
+      status: DocumentStatus.Detached,
+    });
+
+    // 4. When other document is attached, it receives an attached event.
+    const docKey2 = toDocKey(`${task.name}2-${new Date().getTime()}`);
+    const d3 = new yorkie.Document(docKey2);
+    const d4 = new yorkie.Document(docKey2);
+    const eventCollectorD3 = new EventCollector<StatusChangedEvent['value']>();
+    const eventCollectorD4 = new EventCollector<StatusChangedEvent['value']>();
+    const unsub3 = d3.subscribe('status', (event) => {
+      eventCollectorD3.add(event.value);
+    });
+    const unsub4 = d4.subscribe('status', (event) => {
+      eventCollectorD4.add(event.value);
+    });
+    await c1.attach(d3, { syncMode: SyncMode.Manual });
+    await eventCollectorD3.waitAndVerifyNthEvent(1, {
+      status: DocumentStatus.Attached,
+      actorID: c1ID,
+    });
+
+    await c2.activate();
+    c2ID = c2.getID()!;
+    await c2.attach(d4, { syncMode: SyncMode.Manual });
+    await eventCollectorD4.waitAndVerifyNthEvent(1, {
+      status: DocumentStatus.Attached,
+      actorID: c2ID,
+    });
+
+    // 5. When c1 removes a document, it receives a removed event.
+    await c1.remove(d3);
+    await eventCollectorD3.waitAndVerifyNthEvent(2, {
+      status: DocumentStatus.Removed,
+    });
+
+    // 6. When c2 syncs, it should also receive a removed event.
+    await c2.sync();
+    await eventCollectorD4.waitAndVerifyNthEvent(2, {
+      status: DocumentStatus.Removed,
+    });
+
+    // 7. If the document is in the removed state, a detached event should not occur when deactivating.
+    const eventCount3 = eventCollectorD3.getLength();
+    const eventCount4 = eventCollectorD4.getLength();
+    await c1.deactivate();
+    await c2.deactivate();
+    assert.equal(eventCollectorD3.getLength(), eventCount3);
+    assert.equal(eventCollectorD4.getLength(), eventCount4);
+
+    unsub1();
+    unsub2();
+    unsub3();
+    unsub4();
+  });
+
+  it('Should properly handle removed document when deactivating', async function ({
```

Comment:
1. Would you please explain the meaning of “properly handling”? 
2. And it seems that part of validation is written in duplicate with the TC above. is it intended?

Reply:
> 1. Would you please explain the meaning of “properly handling”?

It means that when another peer deactivates a deleted document, the status events occur in the order verified in the test.

> 2. And it seems that part of validation is written in duplicate with the TC above. is it intended?

The event verification is duplicated when first attaching and removing to show which events occur in order. Since the important part is number 3, it might be sufficient to verify only that part.

Reply:
Thanks for your reply. I want some more conversation for clearly understanding.  

1. is peer means client object? 

2. According to your explanation, this tc doesn't need verification with two clients and documents. If this tc's purpose is verification with `c2.deactivate()`, i looks like only needed for c2 and d2. So I think it's enough with just one client and document. Would you please explain what is the point to use c1 and d1?

Reply:
Thank you for your question and for seeking clarification. Let me provide more details:

- The term `peer` is used to refer to a client other than oneself.
- We need c1 and c2 because this test is about a scenario where I deactivate, but another person has deleted the document.

When a client (c1) removes a document, the document's state becomes "removed," and the connection to the server is terminated.

Now, we need to verify how another client (c2) should handle receiving the Change that the document has been removed. This can be examined in two scenarios:

1. When c2 sends a sync request (note that detach behaves similarly as it also receives the change that the document was removed):
   - This is addressed in the test above: `// 6. When c2 syncs, it should also receive a removed event.`

2. When c2 deactivates without syncing (or detaching):
   - This is covered in the current test: `// 3. When c2 deactivates, it should also receive a removed event.`
   - For now, in the case of only deactivating, since c2 doesn't receive the Change that the document was removed, its state changes to "detached."

Reply:
Thank you for your question and for seeking clarification. Let me provide more details:

- The term `peer` is used to refer to a client other than oneself.
- We need c1 and c2 because this test is about a scenario where I deactivate, but another person has deleted the document.

When a client (c1) removes a document, the document's state becomes "removed," and the connection to the server is terminated.

Now, we need to verify how another client (c2) should handle receiving the Change that the document has been removed. This can be examined in two scenarios:

1. When c2 sends a sync request (note that detach behaves similarly as it also receives the change that the document was removed):
   - This is addressed in the test above: `// 6. When c2 syncs, it should also receive a removed event.`

2. When c2 deactivates without syncing (or detaching):
   - This is covered in the current test: `// 3. When c2 deactivates, it should also receive a removed event.`
   - For now, in the case of only deactivating, since c2 doesn't receive the Change that the document was removed, its state changes to "detached."

Reply:
Thanks for your very helpful comment. 
I can almost understand after this comment why this tc was written.

I think this is the point that confused me, I didn't understand why step 2 is needed. It seems like step 2 is created to verifying for removed, but with the current specification, step 3 is checking for detached, and the test succeeds without step 2. (I tested this on my android sdk's test code)

This tc seems to be that contains part of the future spec(number2) and part of the current spec(number3) at the same time, right?

Reply:
@chacha912 
Thanks for your reply.
Now I understand why this TC is contains step 2 and 3.

It's to validate that after calling c1.remove(), the detached event is still being received, and if the spec changes in the future to allow remove to be received, you'll want to modify this TC again then, right?

Reply:
@maruldy Yes, this test (step 3) checks if a 'detached' event occurs when c2 deactivates. And a comment is added to indicate that we might need to modify this test later if we change how things work.

Reply:
Thanks for long conversation. :)  I'll replicate this tc to android sdk. 👍